storage/scylla_api.c

```diff
@@ -0,0 +1,707 @@
+    if (execute_statement(session, statement) != CASS_OK) {
+      ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+      goto exit;
+    }
+  }
+
+exit:
+  cass_prepared_free(insert_prepared);
+  return ret;
+}
+
+status_t select_from_bundleTable(CassSession* session, CassStatement* statement,
+                                 scylla_iota_transaction_t** transaction, size_t* transactionNum) {
+  int idx;
+  status_t ret = SC_OK;
+  CassFuture* future = NULL;
+  const CassResult* result;
+  CassIterator* iterator;
+
+  future = cass_session_execute(session, statement);
+  cass_future_wait(future);
+  if (cass_future_error_code(future) != CASS_OK) {
+    print_error(future);
+    ret = SC_STORAGE_CASSANDRA_QUREY_FAIL;
+    goto exit;
+  }
+
+  /* get transaction number */
+  result = cass_future_get_result(future);
+  iterator = cass_iterator_from_result(result);
+  *transactionNum = 0;
+  while (cass_iterator_next(iterator)) {
+    (*transactionNum)++;
+  }
+  *transaction = malloc((*transactionNum) * sizeof(scylla_iota_transaction_t));
+  if (*transaction == NULL) {
+    ta_log_error("%s\n", "malloc failure");
+    ret = SC_STORAGE_OOM;
+    goto end_iterate;
+  }
+  /* get each transaction result */
+  idx = 0;
+  iterator = cass_iterator_from_result(result);
+  while (cass_iterator_next(iterator)) {
+    const CassRow* row = cass_iterator_get_row(iterator);
+
+    if ((ret = get_blob(row, ((*transaction) + idx)->bundle, "bundle")) != SC_OK) {
+      goto end_iterate;
+    }
+    if ((ret = get_blob(row, ((*transaction) + idx)->address, "address")) != SC_OK) {
+      goto end_iterate;
+    }
+    if ((ret = get_blob(row, ((*transaction) + idx)->hash, "hash")) != SC_OK) {
+      goto end_iterate;
+    }
+    if ((ret = get_blob(row, ((*transaction) + idx)->message, "message")) != SC_OK) {
+      goto end_iterate;
+    }
+    if ((ret = get_blob(row, ((*transaction) + idx)->trunk, "trunk")) != SC_OK) {
+      goto end_iterate;
+    }
+    if ((ret = get_blob(row, ((*transaction) + idx)->branch, "branch")) != SC_OK) {
+      goto end_iterate;
+    }
+    cass_value_get_int64(cass_row_get_column_by_name(row, "value"), &(((*transaction) + idx)->value));
+    cass_value_get_int64(cass_row_get_column_by_name(row, "timestamp"), &(((*transaction) + idx)->timestamp));
+
+    idx++;
+  }
+
+end_iterate:
+  cass_result_free(result);
+  cass_iterator_free(iterator);
+
+exit:
+  cass_future_free(future);
+  cass_statement_free(statement);
+
+  return ret;
+}
+
+static status_t push_columns_into_queue(CassSession* session, CassStatement* statement,
+                                        hash243_queue_t* const res_queue, const char* column_name) {
+  status_t ret = SC_OK;
+  CassFuture* future = NULL;
+  const CassResult* result;
+  CassIterator* iterator;
+
+  future = cass_session_execute(session, statement);
+  cass_future_wait(future);
+
+  if (cass_future_error_code(future) != CASS_OK) {
+    print_error(future);
```

Comment:
use logger instead

Reply:
print_error use ta_logger 
```
static void print_error(CassFuture* future) {
  const char* message;
  size_t message_length;
  cass_future_error_message(future, &message, &message_length);
  ta_log_error("Error: %.*s\n", (int)message_length, message);
}
```
And there's corresponing error message in the caller function.